In [1]:
import pandas as pd 
import numpy as np



C:\Users\OMKAR\anaconda3\lib\importlib\_bootstrap.py:219: RuntimeWarning: numpy.ufunc size changed, may indicate binary incompatibility. Expected 192 from C header, got 216 from PyObject
  return f(*args, **kwds)


In [3]:
data= pd.read_excel("Bank_Personal_Loan_Modelling.xlsx",sheet_name=1);
data.head()

ID  Age  Experience  Income  ZIP Code  Family  CCAvg  Education  Mortgage  \
0   1   25           1      49     91107       4    1.6          1         0   
1   2   45          19      34     90089       3    1.5          1         0   
2   3   39          15      11     94720       1    1.0          1         0   
3   4   35           9     100     94112       1    2.7          2         0   
4   5   35           8      45     91330       4    1.0          2         0   

   Personal Loan  Securities Account  CD Account  Online  CreditCard  
0              0                   1           0       0           0  
1              0                   1           0       0           0  
2              0                   0           0       0           0  
3              0                   0           0       0           0  
4              0                   0           0       0           1

In [4]:
data.isna().sum()

ID                    0
Age                   0
Experience            0
Income                0
ZIP Code              0
Family                0
CCAvg                 0
Education             0
Mortgage              0
Personal Loan         0
Securities Account    0
CD Account            0
Online                0
CreditCard            0
dtype: int64

In [5]:
data.describe()

ID          Age   Experience       Income      ZIP Code  \
count  5000.000000  5000.000000  5000.000000  5000.000000   5000.000000   
mean   2500.500000    45.338400    20.104600    73.774200  93152.503000   
std    1443.520003    11.463166    11.467954    46.033729   2121.852197   
min       1.000000    23.000000    -3.000000     8.000000   9307.000000   
25%    1250.750000    35.000000    10.000000    39.000000  91911.000000   
50%    2500.500000    45.000000    20.000000    64.000000  93437.000000   
75%    3750.250000    55.000000    30.000000    98.000000  94608.000000   
max    5000.000000    67.000000    43.000000   224.000000  96651.000000   

            Family        CCAvg    Education     Mortgage  Personal Loan  \
count  5000.000000  5000.000000  5000.000000  5000.000000    5000.000000   
mean      2.396400     1.937913     1.881000    56.498800       0.096000   
std       1.147663     1.747666     0.839869   101.713802       0.294621   
min       1.000000     0.000000     1.000000     0.000000       0.000000   
25%       1.000000     0.700000     1.000000     0.000000       0.000000   
50%       2.000000     1.500000     2.000000     0.000000       0.000000   
75%       3.000000     2.500000     3.000000   101.000000       0.000000   
max       4.000000    10.000000     3.000000   635.000000       1.000000   

       Securities Account  CD Account       Online   CreditCard  
count         5000.000000  5000.00000  5000.000000  5000.000000  
mean             0.104400     0.06040     0.596800     0.294000  
std              0.305809     0.23825     0.490589     0.455637  
min              0.000000     0.00000     0.000000     0.000000  
25%              0.000000     0.00000     0.000000     0.000000  
50%              0.000000     0.00000     1.000000     0.000000  
75%              0.000000     0.00000     1.000000     1.000000  
max              1.000000     1.00000     1.000000     1.000000

In [6]:
data.dtypes

ID                      int64
Age                     int64
Experience              int64
Income                  int64
ZIP Code                int64
Family                  int64
CCAvg                 float64
Education               int64
Mortgage                int64
Personal Loan           int64
Securities Account      int64
CD Account              int64
Online                  int64
CreditCard              int64
dtype: object

In [11]:
data[data.columns[1:]].corr()['Personal Loan'][:]

Age                  -0.007726
Experience           -0.007413
Income                0.502462
ZIP Code              0.000107
Family                0.061367
CCAvg                 0.366891
Education             0.136722
Mortgage              0.142095
Personal Loan         1.000000
Securities Account    0.021954
CD Account            0.316355
Online                0.006278
CreditCard            0.002802
Name: Personal Loan, dtype: float64

By the correlation we can say that Personal Loan is hightly correlated with income,CCAvg,CDamount

In [25]:
y = data.loc[:,data.columns =="Personal Loan"]
x=data.loc[:, data.columns != 'Personal Loan']

In [26]:
x.head()

ID  Age  Experience  Income  ZIP Code  Family  CCAvg  Education  Mortgage  \
0   1   25           1      49     91107       4    1.6          1         0   
1   2   45          19      34     90089       3    1.5          1         0   
2   3   39          15      11     94720       1    1.0          1         0   
3   4   35           9     100     94112       1    2.7          2         0   
4   5   35           8      45     91330       4    1.0          2         0   

   Securities Account  CD Account  Online  CreditCard  
0                   1           0       0           0  
1                   1           0       0           0  
2                   0           0       0           0  
3                   0           0       0           0  
4                   0           0       0           1

In [27]:
y.head()

Personal Loan
0              0
1              0
2              0
3              0
4              0

In [28]:
import statsmodels.api as sm

In [29]:
x1 = sm.add_constant(x)

In [30]:
Logistic=sm.Logit(y,x1)

In [31]:
result=Logistic.fit()

Optimization terminated successfully.
         Current function value: 0.128331
         Iterations 9


In [32]:
result.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:          Personal Loan   No. Observations:                 5000
Model:                          Logit   Df Residuals:                     4986
Method:                           MLE   Df Model:                           13
Date:                Sun, 09 Aug 2020   Pseudo R-squ.:                  0.5942
Time:                        12:14:56   Log-Likelihood:                -641.65
converged:                       True   LL-Null:                       -1581.0
Covariance Type:            nonrobust   LLR p-value:                     0.000
======================================================================================
                         coef    std err          z      P>|z|      [0.025      0.975]
--------------------------------------------------------------------------------------
const                -11.7179      4.115     -2.847      0.004     -19.784      -3.652
ID                  -5.23e-05   5.14e-05     -1.017      0.309      -0.000    4.85e-05
Age                   -0.0538      0.061     -0.877      0.380      -0.174       0.066
Experience             0.0639      0.061      1.048      0.295      -0.056       0.183
Income                 0.0547      0.003     20.820      0.000       0.050       0.060
ZIP Code           -3.743e-06   4.07e-05     -0.092      0.927   -8.35e-05    7.61e-05
Family                 0.6952      0.074      9.353      0.000       0.549       0.841
CCAvg                  0.1218      0.040      3.070      0.002       0.044       0.200
Education              1.7404      0.115     15.098      0.000       1.514       1.966
Mortgage               0.0005      0.001      0.836      0.403      -0.001       0.002
Securities Account    -0.9453      0.286     -3.305      0.001      -1.506      -0.385
CD Account             3.8232      0.324     11.791      0.000       3.188       4.459
Online                -0.6717      0.157     -4.272      0.000      -0.980      -0.364
CreditCard            -1.1136      0.205     -5.430      0.000      -1.516      -0.712
======================================================================================
"""

The p-values for most of the variables are smaller than 0.05, except five variables, therefore, we will remove them.

In [34]:
df = data.drop(["ID","Age","Experience","ZIP Code","Mortgage"],axis=1)
df.head()

Income  Family  CCAvg  Education  Personal Loan  Securities Account  \
0      49       4    1.6          1              0                   1   
1      34       3    1.5          1              0                   1   
2      11       1    1.0          1              0                   0   
3     100       1    2.7          2              0                   0   
4      45       4    1.0          2              0                   0   

   CD Account  Online  CreditCard  
0           0       0           0  
1           0       0           0  
2           0       0           0  
3           0       0           0  
4           0       0           1

In [38]:
x=df.loc[:, df.columns != 'Personal Loan']
y = df.iloc[:,4:5]
y.head()
# df.describe()

Personal Loan
0              0
1              0
2              0
3              0
4              0

In [39]:
x.head()

Income  Family  CCAvg  Education  Securities Account  CD Account  Online  \
0      49       4    1.6          1                   1           0       0   
1      34       3    1.5          1                   1           0       0   
2      11       1    1.0          1                   0           0       0   
3     100       1    2.7          2                   0           0       0   
4      45       4    1.0          2                   0           0       0   

   CreditCard  
0           0  
1           0  
2           0  
3           0  
4           1

In [41]:
Logistic=sm.Logit(y,x)
result=Logistic.fit()
print(result.summary())

Optimization terminated successfully.
         Current function value: 0.288532
         Iterations 8
                           Logit Regression Results                           
Dep. Variable:          Personal Loan   No. Observations:                 5000
Model:                          Logit   Df Residuals:                     4992
Method:                           MLE   Df Model:                            7
Date:                Sun, 09 Aug 2020   Pseudo R-squ.:                 0.08751
Time:                        12:34:15   Log-Likelihood:                -1442.7
converged:                       True   LL-Null:                       -1581.0
Covariance Type:            nonrobust   LLR p-value:                 5.630e-56
                         coef    std err          z      P>|z|      [0.025      0.975]
--------------------------------------------------------------------------------------
Income                 0.0096      0.001      8.871      0.000       0.008       0.012
Famil

Now All the p-Vlaues are less than 0.05 


We are building the Logistic Regression model

In [45]:
from sklearn.linear_model import LogisticRegression
from sklearn import metrics
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=0)
logreg = LogisticRegression()
logreg.fit(X_train, y_train)

C:\Users\OMKAR\anaconda3\lib\site-packages\sklearn\utils\validation.py:760: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\OMKAR\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='auto', n_jobs=None, penalty='l2',
                   random_state=None, solver='lbfgs', tol=0.0001, verbose=0,
                   warm_start=False)

In [46]:
y_pred = logreg.predict(X_test)
print('Accuracy of logistic regression classifier on test set: {:.2f}'.format(logreg.score(X_test, y_test)))

Accuracy of logistic regression classifier on test set: 0.96


Result is conlusing that this model has 96% of accuracy !!!!

In [47]:
from sklearn.metrics import confusion_matrix
confusion_matrix = confusion_matrix(y_test, y_pred)
print(confusion_matrix)

[[902   8]
 [ 33  57]]


The result is telling us that we have 902+57 correct predictions and 33+8 incorrect predictions.